src/app/signers/SignerDistribution.tsx

```diff
@@ -0,0 +1,501 @@
+        </Text>
+      </Flex>
+      <Text fontSize="sm" fontWeight="semibold" {...rest}>
+        {signerVotingPower.toFixed(2)}%
+      </Text>
+    </Flex>
+  );
+}
+
+function SignersDistributionLegend({
+  signers,
+  onlyShowPublicSigners,
+}: {
+  signers: SignerInfo[];
+  onlyShowPublicSigners: boolean;
+}) {
+  const knownSigners = useMemo(
+    () =>
+      signers
+        .filter(signer => getSignerKeyName(signer.signing_key) !== 'unknown')
+        .map(signer => ({
+          value: signer.weight_percent,
+          name: getSignerKeyName(signer.signing_key),
+        })),
+    [signers]
+  );
+  const unknownSigners = useMemo(
+    () =>
+      signers
+        .filter(signer => getSignerKeyName(signer.signing_key) === 'unknown')
+        .reduce((acc, signer) => acc + signer.weight_percent, 0),
+    [signers]
+  );
+
+  const filiteredSigners = onlyShowPublicSigners
+    ? knownSigners
+    : knownSigners.concat({ name: 'Private signers', value: unknownSigners });
+
+  return (
+    <Stack gap={2}>
+      {filiteredSigners.map(signer => (
+        <SignerLegendItem
+          key={signer.name}
+          signerName={signer.name}
+          signerVotingPower={signer.value}
+        />
+      ))}
+    </Stack>
+  );
+}
+
+const startAngle = 30;
+const endAngle = -(360 - startAngle);
+
+export function SignersDistributionPieChart({
+  signers,
+  onlyShowPublicSigners,
+}: {
+  signers: SignerInfo[];
+  onlyShowPublicSigners: boolean;
+}) {
+  const pieData = useMemo(() => {
+    const knownSignersWithPercentageGreaterThan2 = signers
+      .filter(
+        signer => getSignerKeyName(signer.signing_key) !== 'unknown' && signer.weight_percent > 1
+      )
+      .map(signer => ({
+        name: getSignerKeyName(signer.signing_key),
+        value: signer.weight_percent,
+      }));
+    const knownSignersWithPercentageLessThan2 = signers
+      .filter(
+        signer => getSignerKeyName(signer.signing_key) !== 'unknown' && signer.weight_percent <= 1
+      )
+      .reduce((acc, signer) => acc + signer.weight_percent, 0);
+    const unknownSignersPercentage = signers
+      .filter(signer => getSignerKeyName(signer.signing_key) === 'unknown')
+      .reduce((acc, signer) => acc + signer.weight_percent, 0);
+    return onlyShowPublicSigners
+      ? knownSignersWithPercentageGreaterThan2.concat({
+          name: 'Others',
+          value: knownSignersWithPercentageLessThan2,
+        })
+      : knownSignersWithPercentageGreaterThan2
+          .concat({
+            name: 'Others',
+            value: knownSignersWithPercentageLessThan2,
+          })
+          .concat({
+            name: 'Private signers',
+            value: unknownSignersPercentage,
+          });
+  }, [signers, onlyShowPublicSigners]);
+  const colorMode = useColorMode().colorMode;
+  const theme = useTheme();
+
+  const showLabels = useBreakpointValue({ sm: false, md: false, lg: true, xl: true, '2xl': true });
+
+  const renderActiveShape = useCallback(
```

Comment:
I understand the concern about moving renderActiveShape outside of the component to potentially improve readability and modularity. However, placing the function inside the component keeps related logic together, making it easier to understand the component as a whole. It maintains the context of the function's purpose and usage within the specific component. And it keeps the signature that the Pie component requires vs needing it to become a function that requires the show labels argument in order to conditionally render the labels.

renderActiveShape in its current form is not easily unit testable because it is tightly coupled with the component's internal state and the rendering logic of the recharts library. Testing such a function would typically involve testing the entire rendering process, which is more of an E2E test rather than a unit test.

Reply:
Correct me if I'm wrong but the part from line 247 to 291 has no dependency on component state and just do some math to draw the chart? If that's the case, we would benefit from extracting it as a separate function outside the component?

Reply:
I think it's fair to write unit tests for calculateLabelRadius and the way that I am calculating the inner and outer radii for the shape

Reply:
My point is moving the function outside of the component means I have to change the signature to accept the showlabels state, and then I would have to memoize calling that function with showLabels anyways before passing it to the pie component
function getRenderActiveShape = (showLabels) => {...}
renderActiveShape = (() => getRenderActiveShape(showLabels), [showLabels])
activeShape={renderActiveShape}
or else
activeShape={getRenderActiveShape(showLabels)} which re-renders everytime